### DISCUSSION TITLE
Can it be judged that the cursor is at the beginning of a line

### DISCUSSION THREAD

**Discussion by btea** [nodejs/help#4728] [Q&A] [answered]
Sometimes when I want to use `console.log` to print messages to the terminal, if I use `process.stdout.write` to output content on the terminal elsewhere, the content printed by console.log will be spliced ​​at the end, which does not look very friendly .

At this point, I want to judge whether the current cursor is at the beginning of a line, if so, print the information directly, if not, print `\n` first, and then print the information.

**Comment by tniessen** [comment]
Reciting https://github.com/nodejs/node/issues/48069#issuecomment-1554282618:

> While that's generally possible with many TTY implementations, it will break as soon as `stdout` is not a TTY anymore. Relying on exact cursor placement while mixing different I/O APIs might also lead to race conditions. In general, mixing `console.log()` with `process.stdout.write()` should be avoided whenever possible. There are plenty of npm packages that facilitate the creation of scripts with pretty terminal output. Could you perhaps clarify your use case, as to why you are mixing these APIs for the same stream in the first place?

**Reply by btea** [reply]
Thank you for your detailed reply.

The scenario I encountered was that when packaging with vite, the module information will be output to the terminal through `process.stdout.write()` during module conversion.

But sometimes the plugin may output some warning information through `console.log`. At this time, the output warning information will be spliced ​​behind the output module information, which does not look very clear.

Based on the above scenario, so I opened this discussion.

**Reply by tniessen** [reply]
That seems like an unfortunate design decision within said vite plugin. Logging should be done properly such that the logging framework takes care of formatting the output correctly, or the different output streams should be separated. A single stream should not simultaneously be used both for logging and for other outputs.

**Reply by btea** [reply]
I tried using `process.stdout.clearLine(0)` and `process.stdout.cursorTo(0)` to clear the module information before outputting the warning message. 

However, this will cause the previous module information to be lost, which is not a good solution. 🤔

**Reply by tniessen** [reply, marked as answer]
It will also fail when `process.stdout` is not a TTY, and it will potentially mess up anything written through `process.stdout.write()`. If a part of the application is using `process.stdout` as a non-logging, non-TTY output stream, then trying to simultaneously use it as a TTY will lead to problems in any case.

**Reply by btea** [reply]
Yes, it is true, so the isTTY judgment will be made before use.
